### PR TITLE
[ANOMALY SEGMENTATION] Add multi-metric support to OTE SDK for anomaly tasks

### PR DESCRIPTION
**Pull request by djdameln**
## Description
This PR adds multi-metric support to OTE SDK for anomaly localization tasks (segmentation and detection).

**Changes on OTE SDK:**
- Performance.score is now a property instead of an attribute
- Add `AnomalyLocalizationPerformance` as new subclass of `Performance`
- Add `AnomalyLocalizationScores` class which computes the localization performance from a resultset.
- Add helper functions in `ote_sdk/ote_sdk/utils/dataset_utils` to split dataset into local and globally annotated items (moved from `ote_anomalib`).
- Add `compute_anomaly_localization_scores` to `MetricsHelper`.

**Changes to anomaly task implementation:**
- Remove resultset splitting logic from `evaluate` method and use `compute_anomaly_localization_scores` instead.

> NOTE to the OTE Maintainers
Just a bit of context to @Ilya-Krylov  and @druzhkov-paul, this PR has been created in line with the iAi and UI teams. This is the reason why @atwinand has been added as the reviewer.

## Build
- [BMM](https://ci.iotg.sclab.intel.com/job/IMPT/job/IMPTOps/28142/)

### CODE REVIEW

File: ote_sdk/ote_sdk/entities/metrics.py
Comment (atwinand) [inline ote_sdk/ote_sdk/entities/metrics.py]: Type hints all function signatures please.

File: ote_sdk/ote_sdk/utils/dataset_utils.py
Comment (atwinand) [inline ote_sdk/ote_sdk/utils/dataset_utils.py]: When you do this split, do you change somewhat the original content of the resultset.ground_truth_dataset and resultset.prediction_dataset?

Comment (djdameln) [reply ote_sdk/ote_sdk/utils/dataset_utils.py]: We're splitting the resultset based on what is available in the ground truth dataset. Both the ground truth dataset and the prediction dataset in the original resultset contain a mixture of global and local labels. This causes problems in the metric computation, which always includes all available annotations in the computation. For example, a global (full bbox) annotation will effectively count as a segmentation polygon that masks the entire image when computing the dice score. This is why we need to split the resultset between partially (only global GT annotations) and fully (both local and global GT annotations) annotated items.

The new global dataset will contain all dataset items for which global annotations are available, and the local dataset will contain all dataset items for which local annotations are available. Moreover, the global dataset will only contain the global annotations, and the local dataset will only contain the local annotations. This holds for both the ground truth datasets and the prediction datasets.

To summarize, this is the structure of the original and new resultsets:

**Original resultset**
ground_truth_dataset:
	- global annotations
	- optional local annotations
	
prediction_dataset:
	- global predictions
	- local predictions
	
**New global resultset**
ground_truth_dataset:
	- global annotations
	
prediction_dataset:
	- global predictions
	
**New local resultset**
ground_truth_dataset:
	- local annotations

prediction_dataset:
	- local predictions
	
	
The original resultset will not be modified, because we are creating new dataset items when filling the new datasets. So the original resultset will still have the original dataset items which may contain local annotations in addition to global annotations. This makes sure that we can still use both global and local annotations and predictions when visualizing the results in the UI.

Hope that answers your question?

**Comment by djdameln** [comment]
run ote_sdk tests

File: ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py
Comment (samet-akcay) [inline ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py]: I guess, we'll need another refactor here after merging the detection task. This is because the detection task uses FMeasure. To accommodate this, we would either need another class for detection or `task_type` argument in the constructor to assign the corresponding performance metric.

File: ote_sdk/ote_sdk/utils/dataset_utils.py
Comment (atwinand) [reply ote_sdk/ote_sdk/utils/dataset_utils.py]: Thanks Dick. Quite clear. And also, the AnomalyLocalizationPerformance will then be assigned to the original resultset, correct?

Comment (djdameln) [reply ote_sdk/ote_sdk/utils/dataset_utils.py]: correct

File: ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py
Comment (Ilya-Krylov) [inline ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py:85]: Is it allowed not to have anomalous images in dataset?

Comment (Ilya-Krylov) [reply ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py:85]: It it is allowed, then primary metric can have different kind different from run to run (if the dataset is changed between runs).
If it is not allowed, then it always has local_metric different from None.

File: ote_sdk/ote_sdk/entities/metrics.py
Comment (Ilya-Krylov) [inline ote_sdk/ote_sdk/entities/metrics.py]: Why does it have a name related to Anomaly?
Is it actually needed to have such class instead of using score + dashboard metrics ?

File: ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py
Comment (djdameln) [reply ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py:85]: In an anomaly segmentation project, the user can upload some images but not make any segmentation annotations. In this case the global (image-level label) ground truth dataset will contain anomalous images, but the local (segmentation polygon) ground truth dataset will not contain any anomalous images. When we train the model, we still want to report some performance metric to the user, so we take global score as the primary metric. The user may then annotate some images, after which we can compute the local metric. This will then be the primary metric.

We do this because we always need to report a primary metric to show in the UI. In the future, the UI will show both global and local performance based on the availability. We could then change the AnomalyLocalizationPerformance to always return the local score if needed.

File: ote_sdk/ote_sdk/entities/metrics.py
Comment (djdameln) [reply ote_sdk/ote_sdk/entities/metrics.py]: We need this class to support multi-metric projects. We initially looked at generic MultiScorePerformance implementations, but this is tricky because it would be hard to distinguish the different metrics from each other, especially if not all metrics can be reported at all times (like in the example above with only global, no local annotations). Since anomaly tasks are currently the only tasks that require multi-metric support, we decided to make this class specific to anomaly.

File: ote_sdk/ote_sdk/entities/metrics.py
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/entities/metrics.py:722]: It would be good to implement a proper version of `__repr__` for this class.

File: ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py]: As an option you can combine both global and local dashboard metrics here to provide more information when local score is None.

File: ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py
Comment (samet-akcay) [reply ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py:85]: And a little bit motivation why we do this: We don't want to block the training if there is no detection/segmentation annotation available. We want to train the model as soon as the user uploads normal and abnormal images. If we forced the user to annotate to start the training, the user would be confused as we claim that anomaly tasks are unsupervised tasks.

File: ote_sdk/ote_sdk/utils/dataset_utils.py
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/utils/dataset_utils.py]: Code duplication level is pretty high here. Is it possible to provide some utility functions to get rid of it?

File: ote_sdk/ote_sdk/utils/dataset_utils.py
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/utils/dataset_utils.py]: What about ROIs?

**Review by druzhkov-paul — COMMENTED** [review]
What about tests for the new SDK stuff?

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests

File: ote_sdk/ote_sdk/utils/dataset_utils.py
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/utils/dataset_utils.py]: What about predictions (not ground truth)? I guess, it does not matter for metric evaluation code and these datasets are only created for these purposes. Is it Ok to have `AnnotationSceneKind.ANNOTATION` even for predictions?

File: ote_sdk/ote_sdk/utils/dataset_utils.py
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/utils/dataset_utils.py]: Maybe not the number of annotation?

**Comment by Ilya-Krylov** [comment]
@djdameln could you please resolve conflicts and address remaining comments

**Comment by samet-akcay** [comment]
@djdameln you could maybe point out the issue you're having here.

File: ote_sdk/ote_sdk/entities/metrics.py
Comment (djdameln) [reply ote_sdk/ote_sdk/entities/metrics.py]: I refactored the class so that we now have a generic MultiScorePerformance class that supports adding one or more additional metrics, and a subclass AnomalyLocalizationPerformance which contains the global and local scores specific for anomaly tasks.

**Comment by djdameln** [comment]
run ote_sdk tests

**Comment by djdameln** [comment]
@Ilya-Krylov @druzhkov-paul I addressed your comments. Most importantly, I refactored the new class into a generic `MultiScorePerformance` and an anomaly-specific `AnomalyLocalizationPerformance`, and reduced code duplication in the dataset splitting utils.

One problem that we encountered is the fact that we can't use the resultset purpose property in OTE CLI. Ideally we would like to assign slightly different labels between inference and evaluation. During inference we want to provide a global normal label when no anomalous regions are found. During evaluation we don't want to provide this label, because the segmentation performance is based on a single-class metric. Currently the OTE CLI does not properly assign the purpose in the dataset and resultset classes, so we could not implement this for now. I fixed it temporarily by always adding the normal label, but this will lead to slightly inaccurate local performance metric. I suggest we merge this PR first and then see if we can address the resultset purpose issue. I created a ticket for this: https://jira.devtools.intel.com/browse/CVS-83052

**Comment by druzhkov-paul** [comment]
run ote-test

**Comment by aiarkinx** [comment]
run ote-test

**Comment by djdameln** [comment]
run ote-test

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests

File: external/anomaly/ote_anomalib/callbacks/inference.py
Comment (druzhkov-paul) [inline external/anomaly/ote_anomalib/callbacks/inference.py]: IMO, `append_labels` is a better (right one?) way to do this.

File: external/anomaly/ote_anomalib/callbacks/inference.py
Comment (druzhkov-paul) [inline external/anomaly/ote_anomalib/callbacks/inference.py]: Why can't you just filter that one later during metric calculation?

File: external/anomaly/ote_anomalib/callbacks/inference.py
Comment (druzhkov-paul) [inline external/anomaly/ote_anomalib/callbacks/inference.py]: What does that condition mean? Does it assume that annotation is empty before inference? Why?

File: external/anomaly/ote_anomalib/callbacks/inference.py
Comment (druzhkov-paul) [inline external/anomaly/ote_anomalib/callbacks/inference.py]: Why is it always 0.5?

File: external/anomaly/ote_anomalib/callbacks/inference.py
Comment (djdameln) [reply external/anomaly/ote_anomalib/callbacks/inference.py]: That would also be possible, but since we are using generic helper functions for splitting the dataset in the `evaluate` method, we would still need to check the dataset purpose there.

Comment (druzhkov-paul) [reply external/anomaly/ote_anomalib/callbacks/inference.py]: Can't say that I understood the point. Could you please elaborate a bit, in what cases DatasetItem RoI level normal label makes a problem?

Comment (djdameln) [reply external/anomaly/ote_anomalib/callbacks/inference.py]: > Ideally we would like to assign slightly different labels between inference and evaluation. During inference we want to provide a global normal label when no anomalous regions are found. During evaluation we don't want to provide this label, because the segmentation performance is based on a single-class metric. Currently the OTE CLI does not properly assign the purpose in the dataset and resultset classes, so we could not implement this for now.

As I tried to explain here, having the normal label is problematic when computing the performance metric. The localization performance should be based on the anomalous class only, so any predictions with the `normal` label should be ignored. In the other processing stages (training, inference) however, we do need the normal label. This is why we need to be able to check the dataset/resultset purpose when assigning labels and/or when splitting the dataset into local and global components.

Comment (druzhkov-paul) [reply external/anomaly/ote_anomalib/callbacks/inference.py]: > The localization performance should be based on the anomalous class only, so any predictions with the normal label should be ignored.

This is clear.

> In the other processing stages (training, inference) however, we do need the normal label. 

This is not.

Comment (djdameln) [reply external/anomaly/ote_anomalib/callbacks/inference.py]: During training, we need the normal label so that the model can distinguish between normal and abnormal images. This is mainly because the model computes the adaptive threshold value during validation, which happens on the anomalib side right after training. To compute the adaptive threshold, the model needs to know which images are normal and which are abnormal.

During inference, we need to show the normal label to the user if no anomalous regions were found.

Comment (druzhkov-paul) [reply external/anomaly/ote_anomalib/callbacks/inference.py]: I still miss understanding why can't you remove all normal labels from the predicted dataset right before the call to the metric evaluation function when needed.

Comment (djdameln) [reply external/anomaly/ote_anomalib/callbacks/inference.py]: I see your point. It required some refactoring, but I am now filtering out the normal labels before computing the metrics.

**Comment by djdameln** [comment]
run ote_sdk tests

**Comment by druzhkov-paul** [comment]
run ote_sdk tests

**Comment by druzhkov-paul** [comment]
run ote-test

**Comment by djdameln** [comment]
Did something change in the mypy configuration? Since this morning I'm getting a whole list of mypy errors unrelated to my changes
```
external/anomaly/ote_anomalib/callbacks/progress.py:38: error: Invalid self argument "InferenceParameters" to attribute function "update_progress" with type "Callable[[int], None]"  [misc]
external/anomaly/ote_anomalib/callbacks/progress.py:118: error: Too many arguments  [call-arg]
external/anomaly/ote_anomalib/inference_task.py:97: error: Need type annotation for 'hyper_parameters'  [var-annotated]
external/anomaly/ote_anomalib/inference_task.py:189: error: Incompatible types in assignment (expression has type "AnomalyDetectionScores", variable has type "FMeasure")  [assignment]
external/anomaly/ote_anomalib/inference_task.py:191: error: Incompatible types in assignment (expression has type "AnomalySegmentationScores", variable has type "FMeasure")  [assignment]
external/anomaly/ote_anomalib/openvino.py:134: error: Need type annotation for 'ote_config'  [var-annotated]
external/anomaly/ote_anomalib/openvino.py:154: error: Invalid self argument "InferenceParameters" to attribute function "update_progress" with type "Callable[[int], None]"  [misc]
external/anomaly/ote_anomalib/openvino.py:154: error: Incompatible types in assignment (expression has type "Callable[[], None]", variable has type "Callable[[int], Any]")  [assignment]
external/anomaly/ote_anomalib/openvino.py:231: error: Incompatible types in assignment (expression has type "AnomalyDetectionScores", variable has type "FMeasure")  [assignment]
external/anomaly/ote_anomalib/openvino.py:233: error: Incompatible types in assignment (expression has type "AnomalySegmentationScores", variable has type "FMeasure")  [assignment]
external/anomaly/ote_anomalib/openvino.py:241: error: Need type annotation for 'hparams'  [var-annotated]
external/anomaly/ote_anomalib/openvino.py:287: error: Item "None" of "Optional[ModelEntity]" has no attribute "get_data"  [union-attr]
external/anomaly/ote_anomalib/openvino.py:288: error: Item "None" of "Optional[ModelEntity]" has no attribute "get_data"  [union-attr]
external/anomaly/ote_anomalib/openvino.py:301: error: Invalid self argument "OptimizationParameters" to attribute function "update_progress" with type "Callable[[int], None]"  [misc]
external/anomaly/ote_anomalib/openvino.py:301: error: Too many arguments  [call-arg]
external/anomaly/ote_anomalib/openvino.py:309: error: Invalid self argument "OptimizationParameters" to attribute function "update_progress" with type "Callable[[int], None]"  [misc]
external/anomaly/ote_anomalib/openvino.py:309: error: Too many arguments  [call-arg]
external/anomaly/ote_anomalib/openvino.py:317: error: Item "None" of "Optional[ModelEntity]" has no attribute "get_data"  [union-attr]
external/anomaly/ote_anomalib/openvino.py:318: error: Item "None" of "Optional[ModelEntity]" has no attribute "get_data"  [union-attr]
external/anomaly/ote_anomalib/openvino.py:319: error: Item "None" of "Optional[ModelEntity]" has no attribute "get_data"  [union-attr]
external/anomaly/ote_anomalib/openvino.py:320: error: Item "None" of "Optional[ModelEntity]" has no attribute "get_data"  [union-attr]
external/anomaly/ote_anomalib/openvino.py:330: error: Invalid self argument "OptimizationParameters" to attribute function "update_progress" with type "Callable[[int], None]"  [misc]
external/anomaly/ote_anomalib/openvino.py:330: error: Too many arguments  [call-arg]
external/anomaly/ote_anomalib/nncf_task.py:178: error: Argument "parameters" to "ProgressCallback" has incompatible type "Optional[OptimizationParameters]"; expected "Union[TrainParameters, InferenceParameters, None]"  [arg-type]
Found 24 errors in 4 files (checked 1 source file)
```

**Comment by djdameln** [comment]
run ote_sdk tests

File: external/anomaly/ote_anomalib/openvino.py
Comment (druzhkov-paul) [inline external/anomaly/ote_anomalib/openvino.py]:
```suggestion
                label = self.normal_label if len(annotations) == 0 else self.anomalous_label 
                dataset_item.append_labels([ScoredLabel(label=label, probability=float(probability))])
```
and `dataset_item.append_labels` can be moved to the outside of the if-else clause, because it's the same for all branches.

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests

File: ote_sdk/ote_sdk/utils/dataset_utils.py
Comment (druzhkov-paul) [inline ote_sdk/ote_sdk/utils/dataset_utils.py:54]: This argument is not documented.

**Review by druzhkov-paul — DISMISSED** [review]

**Comment by Ilya-Krylov** [comment]
run ote_sdk tests

File: external/anomaly/ote_anomalib/openvino.py
Comment (djdameln) [reply external/anomaly/ote_anomalib/openvino.py]: Good catch. I've made the changes

**Review by druzhkov-paul — DISMISSED** [review]

**Comment by druzhkov-paul** [comment]
run ote_sdk tests

**Comment by druzhkov-paul** [comment]
run ote-test

**Comment by djdameln** [comment]
run ote_sdk tests

**Comment by djdameln** [comment]
run ote-test

**Review by druzhkov-paul — APPROVED** [review]

**Pull request merged**
